File: backend/src/agents/routers/teams.py
"""POST /api/v1/cells/{cell_id}/teams — provision a team-preset into a cell."""

from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from src._shared.middleware.tenant_context import get_tenant_db_session
from src.agents.models import TeamPreset
from src.agents.schemas import (
    AgentInstanceOut,
    TeamProvisionRequest,
    TeamProvisionResponse,
)
from src.agents.services.team_provisioning_service import TeamProvisioningService
from src.iam.middleware import AuthenticatedUser, get_current_user

router = APIRouter(prefix="/cells/{cell_id}/teams", tags=["agents"])


def get_team_provisioning_service(
    db: AsyncSession = Depends(get_tenant_db_session),
) -> TeamProvisioningService:
    return TeamProvisioningService(db)


@router.post(
    "",
    response_model=TeamProvisionResponse,
    status_code=status.HTTP_201_CREATED,
)
async def provision_team(
    cell_id: UUID,
    payload: TeamProvisionRequest,
    db: AsyncSession = Depends(get_tenant_db_session),
    auth: AuthenticatedUser = Depends(get_current_user),
    service: TeamProvisioningService = Depends(get_team_provisioning_service),
) -> TeamProvisionResponse:
    instances = await service.provision_team(
        preset_slug=payload.preset_key,
        cell_id=cell_id,
        user_id=auth.user.id,
    )
    preset_stmt = select(TeamPreset).where(TeamPreset.slug == payload.preset_key)
    preset = (await db.execute(preset_stmt)).scalar_one()
    return TeamProvisionResponse(
        team_preset_id=preset.id,
        cell_id=cell_id,
        agent_instances=[AgentInstanceOut.model_validate(i) for i in instances],
    )

File: frontend/src/api/teams.ts
/**
 * teams endpoint — provision a team-preset into a cell
 * (POST /cells/{cell_id}/teams — backend/src/agents/routers/teams.py).
 * Schema pinned to `TeamProvisionResponse` (backend/src/agents/schemas.py);
 * reuses `agentInstanceSchema` from api/agents.ts for the nested instances.
 */
import { z } from "zod";
import { apiFetch } from "./client";
import { agentInstanceSchema } from "./agents";

export const teamProvisionResponseSchema = z.object({
  team_preset_id: z.string(),
  cell_id: z.string(),
  agent_instances: z.array(agentInstanceSchema),
});
export type TeamProvisionResponse = z.infer<typeof teamProvisionResponseSchema>;

export interface TeamProvisionPayload {
  /** team_preset.slug to provision — see agents/seed_data/*_v1.py PRESET_SLUG. */
  preset_key: string;
}

export const teamsApi = {
  provision: (cellId: string, payload: TeamProvisionPayload) =>
    apiFetch(`/cells/${cellId}/teams`, {
      method: "POST",
      body: payload,
      schema: teamProvisionResponseSchema,
    }),
};
